clippy_lints/src/manual_arithmetic_check.rs

```diff
@@ -0,0 +1,193 @@
+        && let ExprKind::Binary(op, left, right) = if_block.kind
+        && let BinOpKind::Sub = op.node
+    {
+        if Some(big_var.hir_id) == get_local_var_id(left) && Some(little_var.hir_id) == get_local_var_id(right) {
+            // This part of the condition is voluntarily split from the one before to ensure that
+            // if `snippet_opt` fails, it won't try the next conditions.
+            if let Some(big_var_snippet) = snippet_opt(cx, big_var.span)
+                && let Some(little_var_snippet) = snippet_opt(cx, little_var.span)
+            {
+                span_lint_and_sugg(
+                    cx,
+                    MANUAL_ARITHMETIC_CHECK,
+                    expr_span,
+                    "manual arithmetic check found",
+                    "replace it with",
+                    format!("{big_var_snippet}.saturating_sub({little_var_snippet})"),
+                    Applicability::MachineApplicable,
+                );
+            }
+        } else if Some(little_var.hir_id) == get_local_var_id(left)
+            && Some(big_var.hir_id) == get_local_var_id(right)
+            && let Some(big_var_snippet) = snippet_opt(cx, big_var.span)
+            && let Some(little_var_snippet) = snippet_opt(cx, little_var.span)
+        {
+            span_lint_and_then(
+                cx,
+                MANUAL_ARITHMETIC_CHECK,
+                condition_span,
+                "reverted arithmetic check before subtraction",
+                |diag| {
+                    diag.span_note(
+                        condition_span,
+                        format!("you check that `{big_var_snippet}` is bigger than `{little_var_snippet}`"),
+                    );
+                    diag.span_note(
+                        if_block.span,
+                        format!("but then you subtract `{big_var_snippet}` from `{little_var_snippet}`"),
+                    );
+                    diag.span_suggestion(
+                        if_block.span,
+                        "try replacing it with",
+                        format!("{big_var_snippet} - {little_var_snippet}"),
+                        Applicability::MaybeIncorrect,
+                    );
+                },
+            );
+        }
+    }
+}
+
+fn check_gt(
+    cx: &LateContext<'_>,
+    condition_span: Span,
+    expr_span: Span,
+    big_var: &Expr<'_>,
+    little_var: &Expr<'_>,
+    if_block: &Expr<'_>,
+    else_block: &Expr<'_>,
+) {
+    if let Some(big_var) = Var::new(big_var)
+        && let Some(little_var) = Var::new(little_var)
+    {
+        check_subtraction(cx, condition_span, expr_span, big_var, little_var, if_block, else_block);
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for ManualArithmeticCheck {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let ExprKind::If(condition, if_block, Some(else_block)) = expr.kind
+            && let ExprKind::DropTemps(condition) = condition.kind
```

Comment:
nit: can use [`higher::If`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/higher/struct.If.html), that peels these `DropTemps` for us